R/mcmc-parcoord.R

```diff
@@ -0,0 +1,149 @@
+#' section, below, for details.
+#'
+#' @name MCMC-parcoord
+#' @family MCMC
+#'
+#' @template args-mcmc-x
+#' @template args-pars
+#' @template args-regex_pars
+#' @template args-transformations
+#' @param ... Currently ignored.
+#' @param size,alpha Arguments passed on to \code{\link[ggplot2]{geom_line}}.
+#' @param np For models fit using \code{\link{NUTS}} (more generally,
+#'   any \href{http://en.wikipedia.org/wiki/Symplectic_integrator}{symplectic
+#'   integrator}), an optional data frame providing NUTS
+#'   diagnostic information. The data frame should be the object returned by
+#'   \code{\link{nuts_params}} or one with the same structure.
+#' @param np_style A call to the \code{parcoord_style_np} helper function to
+#'   specify arguments controlling the appearance of superimposed lines
+#'   representing NUTS diagnostics (in this case divergences) if the \code{np}
+#'   argument is specified.
+#'
+#'
+#' @section Plot Descriptions:
+#' \describe{
+#'   \item{\code{mcmc_parcoord}}{
+#'    Parallel coordinates plot of MCMC draws. This can be particularly useful
+#'    if the optional NUTS diagnostic information is provided via the \code{np}
+#'    argument. In that case divergences are highlighted in the plot. This was
+#'    originally suggested by Ari Hartikainen on the Stan Forums and the
+#'    discussion in
+#'    \href{http://discourse.mc-stan.org/t/concentration-of-divergences/1590/21}{that
+#'    thread} may be helpful for understanding the plot.
+#'   }
+#' }
+#'
+#' @examples
+#' color_scheme_set("pink")
+#' x <- example_mcmc_draws(params = 5)
+#' mcmc_parcoord(x)
+#' mcmc_parcoord(x, regex_pars = "beta")
+#'
+#' \dontrun{
+#' library(rstan)
+#' fit <- stan_demo("eight_schools")
+#' draws <- as.array(fit, pars = c("tau", "theta", "lp__"))
+#' np <- nuts_params(fit)
+#'
+#' color_scheme_set("blue")
+#' mcmc_parcoord(draws)
+#' mcmc_parcoord(draws, np = np)
+#'
+#' color_scheme_set("darkgray")
+#' div_style <- parcoord_style_np(div_color = "green", div_size = 0.3, div_alpha = 0.3)
+#' mcmc_parcoord(draws, size = 0.25, alpha = 0.1,
+#'               np = np, np_style = div_style)
+#' }
+NULL
+
+#' @rdname MCMC-parcoord
+#' @export
+mcmc_parcoord <-
+  function(x,
+           pars = character(),
+           regex_pars = character(),
+           transformations = list(),
+           ...,
+           size = 0.2,
+           alpha = 0.3,
+           np = NULL,
+           np_style = parcoord_style_np()) {
+    check_ignored_arguments(...)
+    stopifnot(inherits(np_style, "nuts_style"))
+
+    x <- prepare_mcmc_array(x, pars, regex_pars, transformations)
+    param_labels <- parameter_names(x)
+    draws <- reshape2::melt(merge_chains(x), varnames = c("Iteration", "Parameter"))
+    n_iter <- num_iters(draws)
+    n_param <- num_params(draws)
+
+    if (n_param < 2)
+      stop("This plot requires at least two parameters in 'x'.")
+
+    has_divs <- !is.null(np)
+    if (has_divs) {
+      np <- validate_nuts_data_frame(np)
```

Comment:
Three points.

First, I recommend `dplyr::pull(df, value)` or getElement(df, "value)` instead of `.$value`. I think the package check will be unhappy with the `.` symbol being an undefined global.

Second, speaking of undefined globals, Travis has timed out during checks on this branch. Can you confirm a clean package check?

Third, here's how the function might look in the new tidy evaluation system, you can use `sym()` to create a symbol and then `UQ()` to unquote the symbol. (Typically you would write `!! x` instead of `UQ(x)` but it's best to avoid that shortcut in logical expressions.) Learn more about it from [these guides](http://maraaverick.rbind.io/2017/08/tidyeval-resource-roundup/).

```
    if (has_divs) {
      param <- sym("Parameter")
      value <- sym("Value")
      divg <- sym("Divergent")
      
      draws$Divergent <- np %>% 
        filter(UQ(param) == "divergent__") %>%
        pull(UQ(value)) %>%
        rep(times = n_param)
      
      np <- validate_nuts_data_frame(np)
      div_draws <- filter(draws, UQ(divg) == 1)
      draws <- filter(draws, UQ(divg) == 0)
    }
```

This suggestion is not required. This is just a heads-up to check it out and try to get comfortable with it.